src/resolve/resolved-dns-transaction.c

```diff
@@ -563,28 +625,37 @@ static int dns_transaction_open_tcp(DnsTransaction *t) {
                 return -EAFNOSUPPORT;
         }
 
-        if (fd < 0)
-                return fd;
+        if (!s) {
+                if (fd < 0)
+                        return fd;
 
-        r = dns_stream_new(t->scope->manager, &t->stream, t->scope->protocol, fd);
-        if (r < 0)
-                return r;
-        fd = -1;
+                r = dns_stream_new(t->scope->manager, &s, t->scope->protocol, fd);
+                if (r < 0)
+                        return r;
+
+                fd = -1;
+
+                if (t->server)
+                        s->server = dns_server_ref(t->server);
```

Comment:
dns_server_ref() returns NULL if you pass in NULL (most our _ref() calls do), hence you can drop the if check here, it's redundant